nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/StandardNiFiServiceFacade.java

```diff
@@ -3022,6 +3022,119 @@ public ControllerServiceEntity updateControllerService(final Revision revision,
     }
 
+    @Override
+    public ControllerServiceEntity moveControllerService(final Revision revision, final ControllerServiceDTO controllerServiceDTO, final String newProcessGroupID) {
+        // get the component, ensure we have access to it, and perform the move request
+        final ControllerServiceNode controllerService = controllerServiceDAO.getControllerService(controllerServiceDTO.getId());
+        final RevisionUpdate<ControllerServiceDTO> snapshot = updateComponent(revision,
+                controllerService,
+                () -> moveControllerServiceWork(controllerService, newProcessGroupID),
+                cs -> {
+                    awaitValidationCompletion(cs);
+                    final ControllerServiceDTO dto = dtoFactory.createControllerServiceDto(cs);
+                    final ControllerServiceReference ref = controllerService.getReferences();
+                    final ControllerServiceReferencingComponentsEntity referencingComponentsEntity = createControllerServiceReferencingComponentsEntity(ref);
+                    dto.setReferencingComponents(referencingComponentsEntity.getControllerServiceReferencingComponents());
+                    return dto;
+                });
+
+        final PermissionsDTO permissions = dtoFactory.createPermissionsDto(controllerService);
+        final PermissionsDTO operatePermissions = dtoFactory.createPermissionsDto(new OperationAuthorizable(controllerService));
+        final List<BulletinDTO> bulletins = dtoFactory.createBulletinDtos(bulletinRepository.findBulletinsForSource(controllerServiceDTO.getId()));
+        final List<BulletinEntity> bulletinEntities = bulletins.stream().map(bulletin -> entityFactory.createBulletinEntity(bulletin, permissions.getCanRead())).collect(Collectors.toList());
+        return entityFactory.createControllerServiceEntity(snapshot.getComponent(), dtoFactory.createRevisionDTO(snapshot.getLastModification()), permissions, operatePermissions, bulletinEntities);
+    }
+
+    private ControllerServiceNode moveControllerServiceWork(final ControllerServiceNode controllerService, final String newProcessGroupID) {
+        final ProcessGroup oldParentGroup = controllerService.getProcessGroup();
+        Set<ComponentNode> referencedComponents = controllerService.getReferences().getReferencingComponents();
+        isReferencesDisabled(referencedComponents);
+        oldParentGroup.removeControllerService(controllerService);
+        ProcessGroup newParent;
+        if (!oldParentGroup.isRootGroup() && oldParentGroup.getParent().getIdentifier().equals(newProcessGroupID)) {
+            // move to parent process group
+            newParent = oldParentGroup.getParent();
+            newParent.addControllerService(controllerService);
+
+            // unset any references the controller services has to other controller services that are now out of scope
+            Map<String, String> updatedProps = new HashMap<>();
```

Comment:
Do you need to attempt to remove references to the CS when moving it to the parent process group? Since it is being moved up in the hierarchy, anything that was in scope originally should be in scope still.

Reply:
When moving to the parent group, I'm not checking the references to the CS that is being moved. I'm checking to see if the CS being moved has any references to other CS that may now be out of scope.
For example. CS A and CS B are both in the same group and CS A has a reference to CS B. If I move CS A up one level, CS B is now out of scope, so I have to clear the reference to it on CS A.